Realm/RLMSyncConfiguration.h

```diff
@@ -50,7 +50,40 @@ typedef NS_ENUM(NSUInteger, RLMClientResetMode) {
     /// If RLMClientResetModeDiscardLocal is enabled but the client reset operation is unable to complete
     /// then the client reset process reverts to manual mode. Example) During a destructive schema change this
     /// mode will fail and invoke the manual client reset handler.
-    RLMClientResetModeDiscardLocal
+    ///
+    /// The RLMClientResetModeDiscardLocal mode supports two client reset callbacks -- `RLMClientResetBeforeBlock`, `RLMClientResetAfterBlock` -- which can be passed as arguments when creating the `RLMSyncConfiguration`.
+    ///- see: `RLMClientResetAfterBlock` and `RLMClientResetBeforeBlock`
+    RLMClientResetModeDiscardLocal,
+    /// The client device will download a realm with objects reflecting the latest version of the server. A recovery
+    /// process is run locally in an attempt to integrate the server version with any local changes from before the
+    /// client reset occurred.
+    ///
+    /// The changes are integrated with the following rules:
+    /// 1. Objects created locally that were not synced before client reset will be integrated.
+    /// 2. If an object has been deleted on the server, but was modified on the client, the delete takes precedence and the update is discarded
+    /// 3. If an object was deleted on the client, but not the server, then the client delete instruction is applied.
+    /// 4. In the case of conflicting updates to the same field, the most recent update is applied.
+    ///
+    /// If the recovery process fails, the client reset process falls back to RLMClientResetModeManual.
+    ///
+    /// The RLMClientResetModeRecover mode supports two client reset callbacks -- `RLMClientResetBeforeBlock`, `RLMClientResetAfterBlock` -- which can be passed as arguments when creating the `RLMSyncConfiguration`.
+    ///- see: `RLMClientResetAfterBlock` and `RLMClientResetBeforeBlock`
+    RLMClientResetModeRecover,
+    /// The client device will download a realm with objects reflecting the latest version of the server. A recovery
+    /// process is run locally in an attempt to integrate the server version with any local changes from before the
+    /// client reset occurred.
+    ///
+    /// The changes are integrated with the following rules:
+    /// 1. Objects created locally that were not synced before client reset will be integrated.
+    /// 2. If an object has been deleted on the server, but was modified on the client, the delete takes precedence and the update is discarded
+    /// 3. If an object was deleted on the client, but not the server, then the client delete instruction is applied.
+    /// 4. In the case of conflicting updates to the same field, the most recent update is applied.
+    ///
+    /// If the recovery process fails, the client reset process falls back to RLMClientResetModeDiscardLocal.
```

Comment:
It'd be good to mention *why* it could fail. Is it just if recovery mode is disabled on the server?

Reply:
Yes - that and if an incompatible schema change exists.